### DISCUSSION TITLE
LabGym for Head-Twitch Response

### DISCUSSION THREAD

**Discussion by ako-alt** [umyelab/LabGym#321] [Q&A]
Hello!

Thank you for developing LabGym, I've been using it for the past month or so to develop a model for head-twitch response in mice. It's been super easy to use and figure out, but I'm running into problems with improving the accuracy of detection.

Here is a snippet of what a video looks like:

https://github.com/user-attachments/assets/179d85c9-6333-4e87-9e5d-7e32bbed9b25

The detector that I trained has worked very well, capturing the outline of each mouse and adjusting when the mouse moves. I'm running into problems with the categorizer though. I trained the categorizer on 445 HTR clips and 768 "other" clips, with "other" encompassing movements like quick head turns, grooming, climbing, etc that might get confused for an HTR. I also did a complexity level of 6 for both the pattern recognizer and animation analyzer. I didn't augment any of my prepared training examples when training the categorizer. 

Here were the results I got after training.
| precision | recall | f1-score | support
htr | 0.72 | 0.57 | 0.63 | 3382
other | 0.78 | 0.87 | 0.82 | 5852
accuracy | 0.76 | 0.76 | 0.76 | 0.76
macro avg | 0.75 | 0.72 | 0.73 | 9234
weighted avg | 0.75 | 0.76 | 0.75 | 9234

And once I tested the categorizer, I got these results.
| precision | recall | f1-score | support
htr | 0.91 | 0.83 | 0.87 | 445
other | 0.91 | 0.95 | 0.93 | 768
accuracy | 0.91 | 0.91 | 0.91 | 0.91
macro avg | 0.91 | 0.89 | 0.9 | 1213
weighted avg | 0.91 | 0.91 | 0.91 | 1213

However, when I actually run the "analyze videos" module, I am seeing high undercounting of my HTR, as well as a good number of false positives. Though there are some HTR, especially the more obvious ones, that are captured quite accurately. What should my next step be? Should I add more examples of HTR to the dataset and retrain the categorizer? 

Thanks and let me know if any more information is needed!

**Comment by yujiahu415** [comment]
Hi,

Thanks for the detailed info! I'll help you to improve this. But I need some additional information.

The HTR is fast and subtle, compared to other behaviors. To detect this behavior, you need relatively high resolution and fps of the videos. What is the current video WxH and fps? The clip you showed looks relatively dark. It's not easy to spot the HTR when the ears are not easy to distinguish from the skin color. Maybe you want to increase the brightness of the video a little bit. You can do this in LabGym's preprocessing module. But after you increase the brightness, you may want to test the Detector to see if it's working well. If not, you may need to train a new Detector. You may need to increase the brightness for all your videos.

And what is the duration for each behavior example? And how did you set when generating the behavior examples? Can you share one behavior example pair for me to take a look?

**Reply by ako-alt** [reply]
The video WxH is 640 frames by 480 frames and it's a 30 fps video. I did actually pre-process the clip I pasted above with brightness x 1.5 and contrast kept the same, do you recommend increasing the brightness even more? 

For the behavior samples, I did 10 frames for an animation/pattern image with a skip of 5. I was able to see head-twitches pretty clearly within these behavior samples, though certainly the contrast between the ears and the head isn't super clear. Here is a sample clip and the associated image file. 


https://github.com/user-attachments/assets/fe0c7077-0bc3-4b13-baa8-c4b478331985

![htr pattern image](https://github.com/user-attachments/assets/b2c3c18c-b617-4588-9b45-8a4810578694)


I appreciate your help! Please let me know if there's any more information I can provide.

**Reply by yujiahu415** [reply]
Hi,

The video resolution and fps are fine. The 10-frame duration of behavior example is also appropriate. But you may try increasing the brightness further until the ears can be seen clearly in the behavior examples. This will help the Categorizer to detect the HTR more accurately. And when the ears can be seen clearly, when generating behavior examples, you can try including body parts with STD=50 (or 100, you can test that) in pattern images so that the motion of ears can be displayed in the pattern images. This will further increase the accuracy of the Categorizer to identify this behavior.

After you determine the best setting for generating the behavior examples (you can always discuss the settings with me here and show me some examples so that I can provide suggestions), you need to re-sort all the examples.

For behavior sorting, 445 pairs for HTR and 768 pairs for others are sufficient regarding the amount. One way to increase the accuracy of the Categorizer is to make the 'others' into different categories, like 'head turn', 'grooming', 'climbing', etc. But you don't have to do this for initial sorting. Just stick with the two-category system after you re-generate examples with better settings. Another thing I would like to emphasize is that you need to select the most representative examples for each category. That means, no redundant examples, and the examples you select should cover the majority of different scenarios for a behavior, like the same behavior with different body poses, different orientations, different variants of the same behavior. And always double check with the sorting to eliminate any incorrect sorting, because every example counts. You don't want to confuse the Categorizer with incorrect sorting.

For training settings, complexity level 6 might be too high, you can try using complexity 3 or 4, and input size 32 or 64, for both Animation Analyzer and Pattern Recognizer. You can start from the smaller number and test different combinations. From your training metrics, it seemed you actually used augmentation (from the 'support' metric). You can continue using augmentation (also augment the validation data). You can choose all the augmentation methods.

Let me know how it goes.

**Reply by ako-alt** [reply]
Hi,

I just got back to working on this head-twitch model and just had a few questions. 

I tried different combinations of the Animation Analyzer and Pattern Recognizer as you suggested here, as well as biasing towards AA or PR. I eventually settled on AA LVL 4 Input 64 and PR LVL 4 Input 64, but noticed that the validation loss plateaued around 0.50 while the training loss went down all the way until 0.10, which I believe means that it's overfitting on the training data.

I then increased the PR level to LVL 6, thinking it may need more complexity to capture the temporal dynamics of the head twitch. However, I still have roughly the same validation and training loss. 

I had some ideas -- what do you think would be best for this data set? 

1. Increase the STD from 50 to 100 (or even higher?) and recreate a Categorizer at AA LVL 4 Input 64 and PR LVL 4 Input 64
2. Sort "other" pairs more carefully into different categories like you mentioned
3. Increase the frame duration when sorting behavior examples (ex: I did 10 frames, but maybe doing 15 frames to better capture a total head twitch?)

I wonder if the model is failing because the subtle head twitches could be easily confused for other behaviors like head turns or grooming. Let me know what you think, thanks for the help.

**Reply by yujiahu415** [reply]
LV6 is too complex for this type of behavior. You are right about the feeling of overfitting. So, for an overfitting, the fix is to make the model simpler instead of more complex. You can try lower levels, like 3 or even 2. From my limited understanding, the head-twitch is rather brief, and because it's an important behavior you wanted to identify, I would suggest you not increase the length of one behavior examples to avoid potential mixture of other behaviors into one example. And I would suggest you generate behaviors from one or two videos that have many head-twitches using different length, like 10, 12, and 15 frames, and compare them to see which length gives you the best resolution to distinguish head-twitch from other behaviors like head turns or grooming. You can do the same thing to testing different STD numbers to see which one gives you the best resolution.

At this stage, no need to divide the others to finer categories. Just try out different length and STD to find the best setting. You can paste some examples here if you are unsure about a setting and I can provide feedback. After you settle down the best setting, you can then scale up generating the behavior examples and sort them and your manual sorting effort won't be wasted.

Head twitch is a subtle behavior that is not easy to be identified by computers and can be easily confused with other behaviors. But I'll do my best to help you to find the most suitable setting for it.

Let me know how it goes.

**Reply by ako-alt** [reply]
Hi,

I've played around with the settings and I think I found some good values. I ended up re-creating behavioral examples with a frame duration of 12 and an STD of 25 this time. I trained with AA LVL 3 Input 64 and PR LVL 2 Input 16 (because the head twitch seemed more obvious on the animation than the pattern image). I also only used about 100-150 examples of "htr" and "other." Now, almost every head twitch is tracked and identified in my videos. 

However, a problem has emerged with false positives (for this and models for other behaviors). I want to maximize as many true positives as detected, so is this an inherent drawback that there will be many more false positives? For this head-twitch, I took some of the false positives and sorted them into the "other" category and saw a decrease in accuracy predicting head-twitch. I've also tried to play around with settings (increasing the uncertainty primarily) in the analyze videos module and still notice that there are numerous false positives. 

I also wonder if it's because my frame duration in the initial instance is high (12 frames) but when I'm analyzing I'm going down to 3 frames (which captures the absolute shortest head twitches). Do you have advice for how to reduce false positives while still maximizing the total number of true positives captured in the dataset? I'm ok with allowing some false positives and sorting in post-processing, but unsure how to proceed when there are so many.

Thanks again for all your help!

**Reply by yujiahu415** [reply]
Generally speaking, many false positives mean that the model is too sensitive for detecting the head twitch. To address this, you can remove the ambiguous training examples from the head twitch category to make the categorization boundary more clear-cut. Would you mind sending me (yujiahu415@gmail.com) your sorted examples for me to take a look? If so, I can provide more tailored suggestions.

**Reply by yujiahu415** [reply]
Hi,

Thanks for sending me the sorted dataset. Now I fully understand the dynamics of the htr. The htr is really quick--much quicker than I previously thought, so, 30 fps seems to be a little insufficient to capture the full dynamics of this behavior. But since you already have your videos recorded at 30 fps, there's no room to increase it for now. But in the future, I would suggest you record videos for analyzing htr at 60 fps.

Back to the current videos, I think it's still possible to work this out. You need to reduce the duration of a behavior example. From my observation, the htr mostly happens within 6 frames (0.2 second at 30 fps) of the total 12 frames. I think you can try 6 frames as the duration of behavior examples, instead of 12. You may also generate behavior examples with an interval of 3 frames (skipping 3 frames) to increase the chance of getting perfect examples that are purely htr without mixing with other behaviors, although this might increase the labor of sorting. But I don't think you need too many perfect htr examples, you can probably start with 100 pairs of htr and 100 pairs of others. Quality is much more important than quantity. So, the examples need to be well-selected. Before you start to sort, you may want to send me some examples for a double check. You may also try other values of the duration, like 5 frames or 7 frames, to see which one works the best, but 12-frame is definitely too long.

Other settings like the background and std look good to me. Let me know how it goes. I think you are very close to your analysis goal.
